M2/Macaulay2/m2/enginering.m2

```diff
@@ -225,7 +225,40 @@ commonEngineRingInitializations = (F) -> (
      *-
      )
 
------------------------------------------------------------------------------
+------------------------------------------------------------------------------
+-- TODO improve this, or deprecate promote/lift(List,R,S)
+defaultDegMap := (R,S) -> (
+    n:=degreeLength S-degreeLength R;
+    if n==0 then identity
+    else if n<0 then d -> take(d,degreeLength S)
+    else d -> d|toList(n:0)
+    )
+
+-- automate promote
+setupPromote = method()
+setupPromote (Function,Ring,Ring,Function) := (f,R,S,degmap) -> (
+    promote(R,S) := (a,S) -> f a;
+    promote(List,R,S) := (m,R,S) -> apply(m,degmap);
+    -- promote(Matrix,R,S) :=
+    -- promote(MutableMatrix,R,S) := -- doesn't work, cf https://github.com/Macaulay2/M2/issues/2192
+    promote(Module,R,S) := (M,R1,S1) -> S ** M;
+    promote(Matrix,R,S) := (m,R,S) -> map(promote(target m,S),promote(source m,S),applyTable(entries m,x->promote(x,S)));
+    promote(MutableMatrix,R,S) := (m,R,S) -> mutableMatrix applyTable(entries m,x->promote(x,S));
+    )
+setupPromote (Function,Ring,Ring) := (f,R,S) -> setupPromote(f,R,S,defaultDegMap(R,S));
+
+-- automate (to some extent) lift
+setupLift = method()
+setupLift (Function,Ring,Ring,Function) := (f,R,S,degmap) -> (
+    lift(R,S) := opts -> (a,S) -> if opts.Verify then f a else try f a;
+    lift(List,R,S) := opts -> (m,R,S) -> apply(m,degmap);
+    lift(Module,R,S) := opts -> (M,R,S) -> S ** M;
+    lift(Matrix,R,S) := opts -> (m,R,S) -> map(lift(target m,S),lift(source m,S),applyTable(entries m,x->lift(x,S)));
+    lift(MutableMatrix,R,S) := opts -> (m,R,S) -> mutableMatrix applyTable(entries m,x->lift(x,S));
```

Comment:
Why do these call `lift(x,S)` and not `lift(x, R, S)`?

Also, do we have _any_ `lift` methods of matching e.g. `lift(Module, R, S)` that _isn't_ defined here automatically? Because if not, I think all of these should just be e.g. `lift(Module, Ring, Ring)`, and we can rename `lift(x,S)` to something else that is defined per ring.

Reply:
good question. `lift(Module,...)` only works for free modules anyway, it seems the definition I give could work universally. that being said it's not how it's done for the usual internal `lift`, and I don't dare to change too much.

Reply:
Okay. At the next internals meeting I'll propose deprecating lift/promote for everything other than ring elements. The current system is obfuscated for very little benefit.